#604 Pilot passport challenge

kjporter opened the pull request:
This PR will:
* Implements the Pilot Passport Challenge feature
* Close #604

iccowan commented on app/Http/Controllers/PilotPassportController.php:
```diff
@@ -0,0 +1,193 @@
+        }
+        $p = auth()->guard('realops')->user();
+        $p->delete();
+        $enrollments = PilotPassportEnrollment::where('cid', $request->id);
+        foreach($enrollments as $e) {
+            $e->delete();
+        }
+        $logs = PilotPassportLog::where('cid', $request->id);
+        foreach($logs as $l) {
+            $l->delete();
+        }
```
We shouldn't be using `$request->id` here. We should use `$pilot->cid`. Allowing `$request->id` would allow anyone to maliciously delete anyone's data by injecting some other id

kjporter replied:
Good catch - updated

iccowan commented on app/Http/Controllers/PilotPassportController.php:
```diff
@@ -0,0 +1,193 @@
+    public static function fetchRecentPilotAccomplishments() {
+        $accomplishments = PilotPassportAward::where('awarded_on', '>', now()->subDays(90)->endOfDay())
+            ->orderByRaw('RAND()')->take(10)->get();
+        $ret = null;
+        foreach ($accomplishments as $acc) {
+            $a = (object) [
+                'pilot_name' => $acc->pilot_public_name,
+                'challenge_name' => $acc->challenge_title
+            ];
+            $ret[] = $a;
+        }
+        return $ret;
```
Same here as earlier. I'd prefer for `$ret` or `$a` to not be used, but rather something descriptive. Longer is better than shorter (at the job I used to work, we never had 3 or less letter variables, unless it spelled a full word, but had plenty of 3-4 word variables)

kjporter replied:
Updated

iccowan commented on app/Http/Controllers/PilotPassportController.php:
```diff
@@ -0,0 +1,193 @@
+            return pdf()->html($error_html)->download();
+        }
+        $p = RealopsPilot::find($a->cid);
+        $staff_users = User::with('roles')->where('status', '1')->get();
+        $atm = $staff_users->filter(function ($staff_user) {
+            return $staff_user->hasRole('atm');
+        });
+        if (!$atm) {
+            $atm = $staff_users->filter(function ($staff_user) {
+                return $staff_user->hasRole('datm');
+            });
+        }
+        $atm = $atm->first();
```
We should be able to do this when pulling from the DB to prevent having to do a query and then filter ourselves (and filtering a query is generally faster)

https://laratrust.santigarcor.me/docs/6.x/usage/querying-relationships.html#by-role

iccowan replied:
Ahh okay that's frustrating. For a future story, how difficult would it be to restructure this? Is it like an update or is it an entire delete the User model and start from scratch?

iccowan commented on app/PilotPassport.php:
```diff
@@ -0,0 +1,27 @@
+
+    public static function airfieldInPilotChallenge($airfield, $cid): bool {
+        $enrollments = PilotPassportEnrollment::where('cid', $cid)->get();
+        if ($enrollments->isEmpty()) {
+            return false;
+        }
```
We can take this out. If it's empty, the next loop won't run at all and will return false at the end of the function

kjporter replied:
Good point - done

iccowan commented on app/PilotPassportEnrollment.php:
```diff
@@ -0,0 +1,56 @@
+    protected $table = "pilot_passport_enrollment";
+
+    public function getNameAttribute() {
+        $pilot = RealopsPilot::find($this->cid);
```
I think we should be able to do `$this->pilot()` here since the relationship is defined.

This should be able to be done everywhere with a relationship defined. I'm not gonna comment on all of them (and honestly I'll probably miss a few), but this is one of the benefits with relationships.

kjporter replied:
Done

iccowan commented on database/migrations/2024_09_26_150430_create_pilot_passport_airfield_map_table.php:
```diff
@@ -0,0 +1,34 @@
+            $table->integer('mapped_to');
+        });
+
+        Artisan::call('db:seed', [
+            '--class' => 'PilotPassportAirfieldMapSeeder',
+            '--force' => true
+        ]);
+    }
```
I would prefer to do this via a SQL query, however I kinda like this until the seed fails... Just putting my thoughts down, I don't think anything needs to change?

kjporter replied:
I had to remove this - it causes the browser test to fail because seeding the empty tables is a step.  I figured that I would just manually-issue the command, but maybe that isn't the best idea. How do you suggest that I do this? Check to see if the table is empty and if so, seed it in the update routine?

iccowan replied:
Since this is something that has to exist for the code to work, I would do this with a raw SQL query in the migration. We should be able to use `DB::statement()` or using a transaction if we think that is necessary

iccowan commented on resources/views/emails/pilot_passport/enroll.blade.php:
```diff
@@ -0,0 +1,17 @@
+    our website at <a href="https://www.ztlartcc.org/pilot_passport" alt="Website">https://www.ztlartcc.org/pilot_passport</a>.</p>
+
+<p>Please remember that in order to get credit for visiting a participating airfield, you should plan to land and remain on the ground for 
+    about 5 minutes. You can immediately verify that we logged your visit by refreshing your passport book on our website or by monitoring
+    your email (we will send a confirmation email with each qualifying visit).</p>
```
Maybe including a note that it could be up to 5 minutes after it's been logged that they get the email?

kjporter replied:
It shouldn't be... are you seeing something that would delay the email by 5 minutes?

iccowan replied:
Just with email queuing and if a lot of emails are being sent at a particular time, it could take up to 5 minutes (maybe more if there's really a huge number of emails in the queue) for the email to be processed in that queue

kjporter replied:
Good point - I wasn't thinking about the new email queuing. Updated.

iccowan commented on routes/web.php:
```diff
@@ -48,6 +48,19 @@
+Route::prefix('pilot_passport')->group(function () {
+    Route::get('/', 'PilotPassportController@index')->name('pilotPassportIndex');
+    Route::post('/', 'PilotPassportController@index')->name('pilotPassportIndex');
+    //Route::post('/{pg?}', 'PilotPassportController@index')->name('pilotPassportIndex');
```
Let's remove this comment

kjporter replied:
Done.

iccowan commented on tests/Browser/PilotPassportChallengeTest.php:
```diff
@@ -0,0 +1,145 @@
+        });
+    }
+
+    public function test_change_privacy_settings(): void {
+        $this->browse(function (Browser $browser) {
+            $browser->visit('/pilot_passport')
+                    ->clickLink('Settings')
+                    ->radio('@privacy', '2')
+                    ->press('Save Settings')
+                    ->assertSee('Your privacy preferences have been saved.');
+        });
+    }
```
Can we test that these settings actually get changed?

kjporter replied:
Yes, we can! Done

iccowan commented on tests/Browser/PilotPassportChallengeTest.php:
```diff
@@ -0,0 +1,145 @@
+        });
+    }
+
+    public function test_disenroll_purge(): void {
+        $this->browse(function (Browser $browser) {
+            $browser->visit('/pilot_passport')
+                    ->clickLink('Settings')
+                    ->press('@purge_data')
+                    ->whenAvailable('.modal', function (Browser $modal) {
+                        $modal->type('@confirm', 'confirm - purge all')
+                            ->press('Continue');
+                    })
+                    ->waitFor('.navbar', 30)
+                    ->assertSee('You have been successfully logged out');
+        });
+    }
```
Same here, can we test to make sure the disenroll and purge actually worked? If this logs the user out without actually deleting anything, this would still be successful which would be misleading

kjporter replied:
Sure, easy enough. I added a second test just to check to see if the data has been purged.

iccowan commented on tests/Browser/PilotPassportChallengeTest.php:
```diff
@@ -0,0 +1,145 @@
+            $a->save();
+        }
+    }
+}
```
These are good tests to start and test the frontend, but I think there are some gaps here that would be better suited for unit testing on the backend

kjporter replied:
I'm interested to hear your thoughts on this

iccowan reviewed (changes requested):
Really, really nice work and cool idea!!!

I left a few comments - some changes that should be made and some just thoughts. I'd like to hear your comments before QA...

kjporter commented on app/Console/Commands/PilotPassportActivityUpdate.php:
```diff
@@ -0,0 +1,141 @@
+    // Parameters to consider an aircraft 'at' an airfield
+    private const RADIUS_LIMIT = 3; // NM
+    private const ALTITUDE_LIMIT = 500; // ft
+    private const SPEED_LIMIT = 50; // KTS GS
```
Updated.

kjporter commented on app/Console/Commands/PilotPassportActivityUpdate.php:
```diff
@@ -0,0 +1,141 @@
+        return $data;
+    }
+
+    public static function calcDistance($point1, $point2) {
+        $dist = acos(sin($point1->radLatitude()) * sin($point2->radLatitude()) + cos($point1->radLatitude())
+            * cos($point2->radLatitude()) * cos($point1->radLongitude() - $point2->radLongitude()));
+        return (((180 * 60) / pi()) * $dist);
+    }
```
Done.

kjporter commented on app/Console/Commands/PilotPassportActivityUpdate.php:
```diff
@@ -0,0 +1,141 @@
+                }
+                if (!PilotPassportLog::where('cid', $p->cid)->where('airfield', $a->id)->get()->isEmpty()) {
+                    $this->info('- [Visited Limit] - airfield has already been logged');
+                    break;
```
You're right - this should be a continue so the loop continues checking the next airport. Updated.

kjporter commented on app/Console/Commands/PilotPassportActivityUpdate.php:
```diff
@@ -0,0 +1,141 @@
+                $award->save();
+            }
+        }
+        Mail::to($pilot->email)->send(new PilotPassportMail('phase_complete', $pilot, $c));
+        return true;
```
Restructured to make more sense. I think I had some deprecated code that relied on this to return bool, but that didn't end up in the current draft. Fixed.

kjporter commented on app/Class/LatLng.php:
```diff
@@ -0,0 +1,22 @@
+    public function radLongitude(): float {
+        return $this->longitude * -1 * pi() / 180;
+    }
+}
```
Done

kjporter commented on app/Http/Controllers/PilotPassportController.php:
```diff
@@ -0,0 +1,193 @@
+            $pg = (is_numeric($request->pg)) ? $request->pg : null;
+        }
+        $privacy = null;
+        if (auth()->guard('realops')->user()) {
```
Agreed, but in another story. Now that we're using this for more than once purpose. I considered making a new pilot table, but it makes sense to share.

kjporter commented on app/Http/Controllers/PilotPassportController.php:
```diff
@@ -0,0 +1,193 @@
+        if (!is_numeric($request->challenge_id)) {
+            $valid_enrollment = false;
+        }
+        $p = auth()->guard('realops')->user();
+        if (!$p) {
+            return redirect('/pilot_passport/login');
+        }
```
Both updated.

kjporter commented on app/Http/Controllers/PilotPassportController.php:
```diff
@@ -0,0 +1,193 @@
+        if (!$p) {
+            return redirect('/pilot_passport/login');
+        }
+        $c = PilotPassport::find($request->challenge_id);
```
Done

kjporter commented on app/Http/Controllers/PilotPassportController.php:
```diff
@@ -0,0 +1,193 @@
+        if (!$valid_enrollment) {
+            return redirect(route('pilotPassportIndex'))->withInput(['tab' => 'enrollments'])->with('error', 'Enrollment data invalid. Please contact wm@ztlartcc.org for assistance.');
+        }
+        $e = PilotPassportEnrollment::where('cid', $p->id)->where('challenge_id', $request->challenge_id)->get();
```
Updated where it makes sense and won't cause trouble

kjporter commented on app/Console/Commands/PilotPassportActivityUpdate.php:
```diff
@@ -0,0 +1,141 @@
+                Mail::to($pilot->email)->send(new PilotPassportMail('visited_airfield', $pilot, $a));
+                $this->checkPhaseComplete($pilot);
+                break;
+            }
```
Ok, I moved the inner loop (code that checks if a flight is at an airport) to its own function - that helped clean up the code quite a bit by removing one loop and a break. Let me know what you think.

iccowan replied:
This is a lot cleaner and much easier to read!

iccowan commented on tests/Browser/PilotPassportChallengeTest.php:
```diff
@@ -0,0 +1,159 @@
+        $enrollment = PilotPassportEnrollment::where('cid', $cid)->first();
+        $achievement = PilotPassportAward::where('cid', $cid)->first();
+        $log = PilotPassportLog::where('cid', $cid)->first();
+        if ($pilot || $enrollment || $achievement || $log) {
+            $this->assertTrue(false);
+        }
+        $this->assertTrue(true);
```
We can replace this with:

`$this->assertFalse($pilot || $enrollment || $achievement || $log)`

or if you want to be more descriptive of what's being tested:

```php
$data_remaining = $pilot || $enrollment || $achievement || $log;
$this->assertFalse($data_remaining);
```

 I think I'd also like this to be included in `test_disenroll_purge()` instead of creating a new test. I don't like tests depending on other tests to run for a pass. Especially if we implement parallel testing, we cannot guarantee that tests will run in a specific order every time which could lead to random errors (and tough to debug once we forget about this code)

kjporter replied:
Updated the first test per your comment. Regarding `test_disenroll_purge()` if I add this as another assert, it will give that test 2 asserts which you said is bad practice. This test is specifically checking front-end sequence and I want the assertSee to verify that the user is transported back to the landing page after the logout sequence (assertSee checks for the logout message).

iccowan replied:
Hmm I might go back on what I said there (sorry if that's causing more work). I guess I shouldn't have said one assert, but rather only test one thing (we don't want to be testing more than one component in a test). However, we can't have tests depend on each other because that will eventually cause a weird test failure that we will struggle to track down. I'd rather have 2 associated asserts in a test than 2 separate tests that depend on each other. @c0repwn3r what are your thoughts here?

kjporter replied:
@iccowan Ok, well if this is the case then I need to rebuild most of the tests. I'm a bit confused on this one though, because I feel like we'd be testing the same thing twice (were the database entries purged or not). We'd just need to test for this once, right? The test that you are referencing doesn't ask that question - it is addressing the flow of the UI.

iccowan commented on app/Console/Commands/PilotPassportActivityUpdate.php:
```diff
@@ -0,0 +1,142 @@
+                $award->awarded_on = date('Y-m-d H:i:s');
+                $award->save();
+            }
+            Mail::to($pilot->email)->send(new PilotPassportMail('phase_complete', $pilot, $c));
```
I feel like this should be inside of the above if statement

kjporter replied:
Done. Also added code to ensure that we aren't double-logging the phase complete.

iccowan commented on app/Console/Commands/PilotPassportActivityUpdate.php:
```diff
@@ -0,0 +1,142 @@
+                    break;
+                }
+            }
+            if ($challenge_complete) {
```
Just a nitpick. Can we rename this `$phase_complete`? I think this is talking about the phase and not an individual challenge - just a little confusing as I read through it - please correct me if I'm wrong

kjporter replied:
Good point - done.

iccowan reviewed (commented):
Few more comments... I'm gonna hold off on QA until we get the DB seeding part figured out https://github.com/ZTL-ARTCC/Current_Website/pull/686#discussion_r1827073168

iccowan commented:
QA Bugs:
- [ ] When logged in as a controller and then logging into the pilot passport page, the redirect is to a blank white page
- [ ] The formatting seems to be a little weird on the passport book page. Is this intentional?
<img width="1079" alt="image" src="https://github.com/user-attachments/assets/fe952ebb-bfd6-4fde-966c-186dc7a31e2a">
- [ ] I'm not 100% sure on the best way to QA the process to award achievements. Let me know if there's something I can do here

kjporter commented:
@iccowan both QA bugs that you noted have been resolved:
- Corrected login sequence so it redirects properly when also logged in as a controller
- Fixed initialization of passport book so it will only show 1 book and not multiple
Let me know when you are ready for final QA and I will add an accomplishment so you can test that.

iccowan commented:
I wonder if this may look better as a wide banner that spans the whole bottom of the page rather than in this section. Or if it's in this section, maybe something to take up more of the whitespace.

Maybe it could say "Congratulations to these pilots for recently achieving these milestones!"
And then below it have a 3 column table with the header as the name of achievement and the rows below being names

<table>
  <thead>
    <th>Bug Smasher</th>
    <th>Corporate Pilot</th>
    <th>Airline Captain</th>
  </thead>
  <tbody>
    <td>
      Ian Cowan<br />
      Kyle Porter<br />
      Pilot 3<br />
      Pilot 4<br />
      Pilot 5<br />
    </td>
    <td>
      Ian Cowan<br />
      Kyle Porter<br />
      Pilot 3<br />
      Pilot 4<br />
    </td>
    <td>
      Kyle Porter<br />
      Pilot 5<br />
    </td>
  </tbody>
</table>

iccowan reviewed (approved):
Nicely done! Just have a recommendation above, but not super important. Otherwise, QA looks good

iccowan commented on tests/Browser/PilotPassportChallengeTest.php:
```diff
@@ -0,0 +1,186 @@
+        });
+    }
+
+    public function test_enroll(): void {
+        $this->loginSamplePilot();
+        $this->browse(function (Browser $browser) {
+            $browser->visit('/pilot_passport')
+            ->clickLink('Enrollments')
+            ->press('@enroll_1')
+            ->assertSee('You are now enrolled in the ZTL Pilot Passport program!');
+        });
+    }
```
This is another good spot to check the backend and make sure the pilot was actually enrolled

kjporter replied:
Done!

iccowan commented on tests/Browser/PilotPassportChallengeTest.php:
```diff
@@ -0,0 +1,186 @@
+        });
+    }
+
+    public function test_change_privacy_settings(): void {
+        $this->loginSamplePilot();
+        $this->browse(function (Browser $browser) {
+            $browser->visit('/pilot_passport')
+                    ->clickLink('Settings')
+                    ->radio('@privacy', '2')
+                    ->press('Save Settings')
+                    ->visit('/pilot_passport')
+                    ->clickLink('Settings')
+                    ->assertRadioSelected('@privacy', '2');
+        });
+    }
```
This is another good spot to make sure this has been saved in the backend

kjporter replied:
Done!
